### PR TITLE
J hy python scripts

### PR DESCRIPTION
**Pull request by jbrezmorf**
Scripts for new test scheme.

### CODE REVIEW

File: CMake/Modules/python/PythonLibs.cmake [hunk 1]
```diff
@@ -11,7 +11,6 @@
 find_package(PythonLibs 2.7 REQUIRED)
 
 include_directories(${PYTHON_INCLUDE_DIRS})
-flow_define(HAVE_PYTHON)
```
Comment (jbrezmorf) [inline CMake/Modules/python/PythonLibs.cmake]: Flow still depends on this define. Keep it for now.

File: CMake/Modules/python/PythonScripts.cmake [hunk 2]
```diff
@@ -24,9 +24,21 @@ macro(install_python_lib package_name package_dir)
         message(STATUS "-- cd ${Flow123d_SOURCE_DIR}/CMake/Modules/python && pip install --target=${PYTHON_3RD_PARTY} ${package_name}")
         # we start pip install from directory where setup.cfg is placed thus avoiding issue with --target usage
         # https://github.com/pypa/pip/pull/3450 and https://github.com/pypa/pip/issues/3056
+        # for now we try to install package using this fix (platform centos)
         execute_process(
             COMMAND pip install --target=${PYTHON_3RD_PARTY} ${package_name}
-            WORKING_DIRECTORY ${Flow123d_SOURCE_DIR}/CMake/Modules/python)
+            WORKING_DIRECTORY ${Flow123d_SOURCE_DIR}/CMake/Modules/python
+            RESULT_VARIABLE RETURNCODE)
```
Comment (jbrezmorf) [inline CMake/Modules/python/PythonScripts.cmake]: What is difference here? Pip should not depend on working direcotry.

File: unit_tests/system/python_loader_test.cpp [hunk 3]
```diff
@@ -49,23 +49,29 @@ TEST(PythonLoader, print_error) {
         "Message: global name 'a' is not defined\nTraceback");
 }
 
+
```
Comment (jbrezmorf) [inline unit_tests/system/python_loader_test.cpp]: We should allow build and run of flow only with packages that are absolutely necessary.

File: bin/python/exec_parallel.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    '(using platform.node() or socket.gethostname()) and then search',
+    'it in the table "host_table.json" which assign logical hostname',
+    'possibly to multiple different real hostnames. ',
+    '',
+    'If the real host name is not found in the table ',
+    'it is used directly otherwise the logical ',
+    'hostname is used to select the setup script.',
+])
+parser.add('-v', '--valgrind', type=[True, str], name='valgrind', placeholder='[<VALGRIND ARGS>]', docs=[
+    'Run tests under valgrind, with python suppression with optional argument',
```
Comment (jbrezmorf) [inline bin/python/exec_parallel.py]: Not only tests.

File: bin/python/exec_parallel.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    ' <valgrind args> passed to the valgrind. (In PBS mode this arguments is ignored.)?',
+])
+parser.add('', '--batch', type=True, name='batch', docs=[
+    'Make output of this script more for an off-line reading',
```
Comment (jbrezmorf) [inline bin/python/exec_parallel.py]: I do not understand the doc.

File: bin/python/exec_parallel.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+parser.add_section('Passable arguments to exec_with_limit.py')
+parser.add('-t', '--limit-time', type=Duration.parse, name='time_limit', placeholder='<time>', docs=[
+    'Obligatory wall clock time limit for execution in seconds or in HH:MM:SS format.',
+    'For precision use float value'
+])
+parser.add('-m', '--limit-memory', type=float, name='memory_limit', placeholder='<memory>', docs=[
+    'Optional memory limit per node in MB',
+    'For precision use float value'
+])
+parser.add('', '--root', hidden=True, type=str, name='root', placeholder='<ROOT>', docs=[
```
Comment (jbrezmorf) [inline bin/python/exec_parallel.py]: Is it necessary if this tool is for any executable?

File: bin/python/host_table.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (jbrezmorf) [inline bin/python/host_table.json]: We should move this into config dir. Possibly use YAML format namely to add comment with explanation of purpose of this file.

File: src/python/scripts/core/base.py [hunk 8]
```diff
@@ -0,0 +1,463 @@
+#!/usr/bin/python
```
Comment (jbrezmorf) [inline src/python/scripts/core/base.py]: For all python sources: document at least purpose of every class.

File: src/python/scripts/exec_parallel_module.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+        template = PBSModule.format(
+            module.template,
+            command=command,
+            json_output=case.fs.json_output # TODO remove
+        )
+
+        return template
+
+    def run_local_mode_one(self, proc):
```
Comment (jbrezmorf) [inline src/python/scripts/exec_parallel_module.py]: document difference against run_local_mode

File: src/python/scripts/pbs/modules/pbs_hydra_kai_tul_cz.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    def __init__(self, job_id, case):
+        super(ModuleJob, self).__init__(job_id, case)
+        self.parser = self.parser_builder(
+            self, 4, JobState.UNKNOWN,
+            queue=7,
+            name=2,
+        )
+
+    @classmethod
+    def update_command(cls):
```
Comment (jbrezmorf) [inline src/python/scripts/pbs/modules/pbs_hydra_kai_tul_cz.py]: Seems that interface classes  to various PBS systems only define commands however parsing of the output should be also specific.

**Comment by jbrezmorf** [comment]
TODO:
1. merge master
2. add doc for classes
3. Is comparison finished?
4. Some test of PBS specific  interface. So that we can run it on a particular parallel machine and be shure that it works. Need to test:
- can execute serial  job (np 1) under PBS
- can execute parallel job (np 2) under PBS
- can handle error in one process of the parallel job
- can handle job rejected by PBS (e.g. wrong queue)
- can handle job waiting to long for execution (How we treat this case?) 

Further discussion necessary.

File: CMake/Modules/python/PythonLibs.cmake (hunk 1 above)
Comment (x3mSpeedy) [inline CMake/Modules/python/PythonLibs.cmake]: `flow_define(HAVE_PYTHON)` is still present in `CMakeLists.txt` no change required

File: CMake/Modules/python/PythonScripts.cmake (hunk 2 above)
Comment (x3mSpeedy) [inline CMake/Modules/python/PythonScripts.cmake]: Its due to bug. In working directory is present special configuration file which `pip` recognize. Other solution is to copy this file to current working directory.

File: unit_tests/system/python_loader_test.cpp (hunk 3 above)
Comment (x3mSpeedy) [inline unit_tests/system/python_loader_test.cpp]: Do you mean during Python copy? Packed Python packages folder has `21 MB` I don't see any big benefit by reducing size even more. Further discussion may be required on this topic.

File: bin/python/exec_parallel.py (hunk 6 above)
Comment (x3mSpeedy) [inline bin/python/exec_parallel.py]: It is for any executable but host_table.json is needed in fixed location. But will remove.

File: bin/python/exec_parallel.py (hunk 4 above)
Comment (x3mSpeedy) [inline bin/python/exec_parallel.py]: Fixed

File: bin/python/exec_parallel.py (hunk 5 above)
Comment (x3mSpeedy) [inline bin/python/exec_parallel.py]: Add more notes on this flag

File: unit_tests/system/python_loader_test.cpp [hunk 11]
```diff
@@ -49,23 +49,29 @@ TEST(PythonLoader, print_error) {
...
+// only test embedded python if we actually copied out python
```
Comment (jbrezmorf) [inline unit_tests/system/python_loader_test.cpp]: Change comment to Test that necessary python libraries are available in embedded python.

**Pull request merged**